### PR TITLE
Fix index for datetime64 conversion. Fixes #13937

### LINKED ISSUE #13937: BUG: DataFrame.to_records() bug in converting datetime64 index with timezone
#### Fix

in to_records(), use pandas.core.common.is_datetime64_any_dtype instead of pandas.core.common.is_datetime64_dtype to check to see if the index is in fact of the datetime64 type.
#### Code Sample, a copy-pastable example if possible

```
import datetime
import pandas as pd
import pytz

data = [datetime.datetime.now(pytz.timezone('UTC')) for _ in range(10)]
df = pd.DataFrame({'datetime': data})
df.set_index('datetime', inplace=True)
df.to_records()
```
#### Expected Output

data
## INSTALLED VERSIONS

commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 20.3
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.2
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0
pandas_datareader: None

### PR DESCRIPTION
**Pull request by amolkahat**
- [x] closes #13937
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] 0.20.0

Fix index for datetime64 conversion

This PR fixes index for datetime64 conversion. Also,
update tests - pandas/pandas/tests/tests_convert_to.py

Fixes https://github.com/pydata/pandas/issues/13937

### CODE REVIEW

**Comment by amolkahat** [comment]
@jreback Please take a look. I updated tests as you suggested in PR #14044.

File: pandas/tests/frame/test_convert_to.py [hunk 1]
```diff
@@ -175,7 +172,34 @@ def test_to_records_index_name(self):
     def test_to_records_with_unicode_index(self):
         # GH13172
         # unicode_literals conflict with to_records
-        result = DataFrame([{u'a': u'x', u'b': 'y'}]).set_index(u'a')\
+        result = DataFrame([{u'a': u'x', u'b': 'y'}]).set_index(u'a') \
             .to_records()
         expected = np.rec.array([('x', 'y')], dtype=[('a', 'O'), ('b', 'O')])
         tm.assert_almost_equal(result, expected)
+
+    def test_to_records_with_tz_gmt(self):
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: can you add a comment for the issue number

**Comment by jreback** [comment]
pls add a whatsnew entry for 0.19.1

File: pandas/tests/frame/test_convert_to.py [hunk 2]
```diff
@@ -175,7 +172,71 @@ def test_to_records_index_name(self):
...
+        """
+        This test verifies the bug id #13937
+        """
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_convert_to.py]: Can you add this as a comment instead of docstring? (docstrings hide the test name in the output of nosetests)

Also just the number is fine, like `# GH 13937 `

File: pandas/tests/frame/test_convert_to.py [hunk 3]
```diff
@@ -175,7 +172,71 @@ def test_to_records_index_name(self):
...
+               for i in range(10)]
+        dt3 = [pytz.utc.localize(dt1[i]) for i in range(10)]
+
+        df1 = DataFrame({'datetime': dt2})
+        df2 = df1.set_index('datetime')
+
+        df3 = DataFrame({'datetime': dt3})
+        df4 = df3.set_index('datetime')
+
+        tm.assert_frame_equal(df2, df4)
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_convert_to.py]: Is this test related to the bug you fixed? 
It does not use `to_records`

File: pandas/tests/frame/test_convert_to.py [hunk 4]
```diff
@@ -175,7 +172,71 @@ def test_to_records_index_name(self):
...
+                                 pytz.UTC) for i in range(10)]
+
+        df1 = DataFrame({'datetime': dt1})
+        df2 = df1.set_index('datetime')
+        df3 = df2.tz_convert("UTC")
+
+        df4 = DataFrame({'datetime': dt2})
+        df5 = df4.set_index('datetime')
+
+        tm.assert_frame_equal(df3, df5)
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_convert_to.py]: Same question for this test.
Was the above code failing before your PR? Or is this not yet tested?

File: pandas/tests/frame/test_convert_to.py [hunk 5]
```diff
@@ -175,7 +172,34 @@ def test_to_records_index_name(self):
...
             .to_records()
         expected = np.rec.array([('x', 'y')], dtype=[('a', 'O'), ('b', 'O')])
         tm.assert_almost_equal(result, expected)
+
+    def test_to_records_with_tz_gmt(self):
+        data = [datetime.datetime.now(pytz.timezone('UTC')) for _ in range(10)]
+        df = DataFrame({'datetime': data})
+        df.set_index('datetime', inplace=True)
+        dfgmt = df.tz_convert("GMT")
+        tm.assertIsNot(df.to_records()['datetime'][0].tzinfo,
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_convert_to.py]: Can you also simplify the construction of the test data using pandas constructs?

Eg 

```
data = [datetime.datetime.now(pytz.timezone('UTC')) for _ in range(10)]
df = DataFrame({'datetime': data})
df.set_index('datetime', inplace=True)
```

can be done with (for similar object, not exact same values of course)

```
df = DataFrame(index=date_range('2016-01-01', periods=10, freq='S', tz='UTC'))
```

File: pandas/tests/frame/test_convert_to.py (hunk 3 above)
Comment (amolkahat) [inline pandas/tests/frame/test_convert_to.py]: @jorisvandenbossche Hmm. yes it is not related to my bug fix. I just focused on data frame equality. Actually there is also one problem assert_frame_equal does not accept any other frames expect DataFrame. In my previous PR 14044 Jeff suggest to construct equal data frame. Can you suggest me how i complete Jeff's assert_frame_equal requirement?

File: pandas/tests/frame/test_convert_to.py (hunk 4 above)
Comment (amolkahat) [inline pandas/tests/frame/test_convert_to.py]: No this code does not falling before my PR or after the fixes.

File: pandas/tests/frame/test_convert_to.py (hunk 3 above)
Comment (amolkahat) [inline pandas/tests/frame/test_convert_to.py]: @jorisvandenbossche @jreback Can you please suggest me how i construct equal data frame?

Comment (jorisvandenbossche) [inline pandas/tests/frame/test_convert_to.py]: Your question is not fully clear to me. 
If you want to test `to_records`, you can also assert that the resulting array is equal (with `assert_numpy_array_equal` instead of `assert_frame_equal`)

**Comment by jorisvandenbossche** [comment]
@amolkahat It looks like something went wrong with rebasing/updating. Normally, doing exactly the following should clean it up:

```
git fetch upstream
git checkout bug_fixes
git rebase upstream/master
git push origin bug_fixes -f
```

(supposing your fork is 'origin' and pandas-dev is 'upstream')

**Comment by jorisvandenbossche** [comment]
@amolkahat Can you rebase? (see my previous comment how to fix it)

**Comment by amolkahat** [comment]
@jorisvandenbossche Please take a look.

**Comment by jreback** [comment]
can you add a whatsnew note for 0.20.0

File: pandas/tests/frame/test_convert_to.py [hunk 6]
```diff
@@ -175,7 +171,108 @@ def test_to_records_index_name(self):
...
+        df = DataFrame({'datetime': date_range('2016-01-01',
+                                               periods=10,
+                                               freq='S',
+                                               tz='UTC')},
+                       index=date_range('2016-01-01',
+                                        periods=10,
+                                        freq='S',
+                                        tz='UTC'))
+
+        dfgmt = df.tz_convert("GMT")
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: use ``result = ....``

can you construct a rec-array directly and use ``tm.assert_numpy_array_equal``

File: pandas/tests/frame/test_convert_to.py [hunk 7]
```diff
@@ -175,7 +171,108 @@ def test_to_records_index_name(self):
...
+                                   freq='S',
+                                   tz='GMT'),
+                        index=dt3)
+
+        tm.assert_numpy_array_equal(df1.to_records(), df3.to_records())
+
+    def test_from_UTC_to_GMT_as_equal_frame(self):
+        # GH13937
+
+        dt1 = [datetime.datetime(2011,
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: this last test is not necessary

File: pandas/tests/frame/test_convert_to.py [hunk 8]
```diff
@@ -175,7 +171,108 @@ def test_to_records_index_name(self):
...
+                                 8,
+                                 15,
+                                 12,
+                                 i + 1,
+                                 pytz.UTC) for i in range(10)]
+
+        df1 = DataFrame(index=dt1)
+        df3 = df1.tz_convert("UTC")
+
+        df4 = DataFrame(index=dt2)
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: however the empty frame is good (but instead of a separate tests, you can do this as part of the other one, IOW)
just use a starting from from the first test

File: pandas/tests/frame/test_convert_to.py [hunk 9]
```diff
@@ -175,7 +171,108 @@ def test_to_records_index_name(self):
...
+                                      freq='S',
+                                      tz='GMT'),
+                           index=date_range('2016-01-01',
+                                            periods=10,
+                                            freq='S',
+                                            tz='UTC'))
+        dfgmt = df_gmt[:].tz_convert("UTC")
+        tm.assert_numpy_array_equal(df_utc.to_records(), dfgmt.to_records())
+
+    def test_to_records_with_tz_as_equal_frame(self):
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: try to consolidate these tests into a single test (but with multiple parts is ok), its just confusing to start with 3 different frames, start with 1, then do a 

```
result = ....
expected = ....
tm.assert_numpy_array_equal(....)

# next part
....

# last part
...
```

**Review by jreback — CHANGES_REQUESTED** [review]
try to consolidate these tests into a single test (but with multiple parts is ok), its just confusing to start with 3 different frames, start with 1, then do a 

```
result = ....
expected = ....
tm.assert_numpy_array_equal(....)

# next part
....

# last part
...
```

File: doc/source/whatsnew/v0.20.0.txt [hunk 10]
```diff
@@ -355,3 +355,7 @@ Bug Fixes
 
 
 - Bug in ``pd.read_csv()`` for the C engine where ``usecols`` were being indexed incorrectly with ``parse_dates`` (:issue:`14792`)
+=======
+- Bug in ``DataFrame.to_records()`` with converting datetime64 index with timezone (:issue: `13937`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: so looks like some merge conflicts are remaining here

File: pandas/tests/frame/test_convert_to.py [hunk 11]
```diff
@@ -175,7 +172,117 @@ def test_to_records_index_name(self):
...
+                                    df_gmt2.to_records())
+
+    def test_from_UTC_to_GMT_as_equal_frame(self):
+        # GH13937
+
+        data1 = [datetime.datetime(2011,
+                                   8,
+                                   15,
+                                   8,
+                                   15,
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: this is really hard to read. please format better.

File: pandas/tests/frame/test_convert_to.py [hunk 12]
```diff
@@ -175,7 +172,117 @@ def test_to_records_index_name(self):
...
+        df_gmt = DataFrame({'datetime': date_range('2016-01-01',
+                                                   periods=10,
+                                                   freq='S',
+                                                   tz='GMT')},
+                           index=date_range('2016-01-01',
+                                            periods=10,
+                                            freq='S',
+                                            tz='GMT'))
+
+        tm.assert_almost_equal(df_utc.to_records()['datetime'][0],
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: what does this comparison add that the ``assert_numpy_array_equal`` does not? (which is the way we would like to compare these)

File: pandas/tests/frame/test_convert_to.py [hunk 13]
```diff
@@ -175,7 +172,117 @@ def test_to_records_index_name(self):
...
+                 for i in range(10)]
+
+        df_utc2 = DataFrame(date_range('2016-01-01',
+                                       periods=10,
+                                       freq='S',
+                                       tz='UTC'),
+                            index=data2)
+
+        df_gmt2 = DataFrame(date_range('2016-01-01',
+                                       periods=10,
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: what is this test actually testing? (which the first test does not)

**Review by jreback — CHANGES_REQUESTED** [review]
not sure if you updated this. the tests needs to be more concise and focused.

**Comment by jreback** [comment]
can you update

**Comment by amolkahat** [comment]
@jreback Please take a look.

File: doc/source/whatsnew/v0.20.0.txt [hunk 14]
```diff
@@ -589,3 +589,5 @@ Bug Fixes
 - Bug in ``Series.replace`` and ``DataFrame.replace`` which failed on empty replacement dicts (:issue:`15289`)
 - Bug in ``pd.melt()`` where passing a tuple value for ``value_vars`` caused a ``TypeError`` (:issue:`15348`)
 - Bug in ``.eval()`` which caused multiline evals to fail with local variables not on the first line (:issue:`15342`)
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: move into a blank space above so that you don't get conflicts.

``with converting a datetime64 index``

File: pandas/tests/frame/test_convert_to.py [hunk 15]
```diff
@@ -2,19 +2,19 @@
...
+from numpy import nan
 
 import pandas.util.testing as tm
-
+from pandas import DataFrame, MultiIndex, Series, Timestamp, compat, date_range
 from pandas.tests.frame.common import TestData
 
 
 class TestDataFrameConvertTo(tm.TestCase, TestData):
+    _multiprocess_can_split_ = True
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: remove as we are using pytest now

File: pandas/tests/frame/test_convert_to.py [hunk 16]
```diff
@@ -173,7 +173,79 @@ def test_to_records_index_name(self):
...
+        # GH13937
+        df_utc = DataFrame({'datetime': date_range('2016-01-01',
+                                                   periods=10,
+                                                   freq='S',
+                                                   tz='UTC')},
+                           index=date_range('2016-01-01',
+                                            periods=10,
+                                            freq='S',
+                                            tz='UTC'))
+
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: do something simpler and much more readable

```
result = df_utc.tz_convert("GMT").to_records()
expected = df_utc.to_records()

tm.assert_numpy_array_equal(result, expected)
```

File: pandas/tests/frame/test_convert_to.py [hunk 17]
```diff
@@ -173,7 +173,79 @@ def test_to_records_index_name(self):
...
         # unicode_literals conflict with to_records
-        result = DataFrame([{u'a': u'x', u'b': 'y'}]).set_index(u'a')\
+        result = DataFrame([{u'a': u'x', u'b': 'y'}]).set_index(u'a') \
             .to_records()
         expected = np.rec.array([('x', 'y')], dtype=[('a', 'O'), ('b', 'O')])
         tm.assert_almost_equal(result, expected)
+
+    def test_to_records_with_tz(self):
+        # GH13937
+        df_utc = DataFrame({'datetime': date_range('2016-01-01',
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: can you format this to make it easier to read

File: pandas/tests/frame/test_convert_to.py [hunk 18]
```diff
@@ -173,7 +173,79 @@ def test_to_records_index_name(self):
...
+
+        tm.assertIsNot(df_utc.to_records().index[0].tzinfo,
+                       df_utc.tz_convert("GMT").to_records().index[0].tzinfo)
+
+        tm.assert_numpy_array_equal(df_utc.to_records(),
+                                    df_utc.tz_convert("GMT").to_records())
+
+        # test DataFrame.to_records() with timezone conversion to UTC
+
+        df_gmt = DataFrame({'datetime': date_range('2016-01-01',
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: same with all of the following, make shorter / simpler / easier-to-read

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@amolkahat thanks it looks a bit better. In general, please don't modify whitespace. You must have some editor / settings which are doing things. They are pretty distracting.

**Comment by jreback** [comment]
you'll want to rebase on master as travis was having some issues

**Comment by jreback** [comment]
thanks!

**Comment by amolkahat** [comment]
@jreback Thanks.
Can you tell me on which IRC server & channel pandas-dev discussion happen?

**Comment by jorisvandenbossche** [comment]
@amolkahat for chat, there is the gitter chat (https://gitter.im/pydata/pandas), there is also the pandas-dev mailing list (https://mail.python.org/mailman/listinfo/pandas-dev), but most of the discussion happens on github

**Comment by amolkahat** [comment]
@jorisvandenbossche Thank you. 😃
